docs/design/remerge/rfc.md

```diff
@@ -0,0 +1,938 @@
+          duplicate record's id to the new ID, and proceed as if we had a
+          local record with this ID (and a mirror with it, if the duplicate
+          had a mirror record).
+
+    3. If the incoming record is not in conflict with our local record (see
+       the section on vector clocks for how we determine conflict), then we
+       forward the mirror and local records to the incoming record.
+
+    4. If the incoming record *is* in conflict with our local record, then we
+       take one of the following steps:
+
+        1. If the local record is a tombstone and the collection is set to
+           prefer deletions, we resolve in favor of the local tombstone.
+
+            - Note that this still involves the conflict check -- tombstones
+              still have clocks.
+
+        2. If the local record is a tombstone and the collection is set to
+           prefer updates, we forward the local and mirror to the incoming
+           remote record.
+
+        3. If we don't have a mirror record, we need to perform a two way merge.
+
+        4. If we have a mirror record, and the incoming record is a descendent
+           of the mirror, then we can perform a three way merge correctly. This
+           is likely to be the most common case.
+
+        5. If we have a mirror record, and the incoming record is in conflict
+           with the mirror record, then we should either
+
+            1. For non-legacy collections, discard the record.
+
+                - This should never happen unless the client that wrote the record
+                  has a bug. Otherwise, it would perform a merge with server before
+                  uploading the record.
+
+                - Note that we wipe the mirror when node reassignment/password reset
+                  occur, so that case doesn't apply here.
+
+            2. For legacy collections, which could have had the vector clock wiped
+               by a legacy client, assume this is what has happened, and do a three
+               way merge.
+
+    5. If we performed a two or three way merge, and the outcome of the merge
+       indicates that we must duplicate the record, then
+
+        1. We create a new local record with identical contents (but a new ID, and
+           fresh vector clock) to the current local version.
+
+        2. Then replacing the current local and mirror with the incoming record.
+
+    6. All new and locally changed records (including records that were not
+       resolved by forwarding the local and mirror versions to the incoming
+       record) must be uploaded as part of this sync.
+
+        - Before uploading the records, we validate them against the schema. If
+          validation fails, we record telemetry, and only sync the records which
+          pass validation.
+            - Records which aren't synced should be flagged as such (XXX: flesh this out)
+            - This telemetry should be monitored, as it indicates a bug.
```

Comment:
Oh, I meant if we only sync records that pass validation, and leave the ones that fail unsynced, instead of failing the entire sync. If one of the records that failed to validate is referenced by another, we'll end up uploading incomplete data.